diffrax/_step_size_controller/jump_step_wrapper.py

```diff
@@ -0,0 +1,410 @@
+            step_inc_cond = keep_step & (t1 == _get_t(i_step, step_ts))
+            i_step = jnp.where(step_inc_cond, i_step + 1, i_step)
+
+        if jump_ts is not None:
+            next_jump_t = _get_t(i_jump, jump_ts)
+            jump_inc_cond = keep_step & (t1 >= eqxi.prevbefore(next_jump_t))
+            i_jump = jnp.where(jump_inc_cond, i_jump + 1, i_jump)
+
+        if self.rejected_step_buffer_len > 0:
+            assert rjct_buff is not None
+            # If the step ended at t1==rjct_buff[i_rjct], then we have successfully
+            # stepped to this time and we increment i_rjct.
+            # We increment i_rjct even if the step was rejected, because we will
+            # re-add the rejected time to the buffer immediately.
+            rjct_inc_cond = t1 == _get_t(i_rjct, rjct_buff)
+            i_rjct = jnp.where(rjct_inc_cond, i_rjct + 1, i_rjct)
+
+            # If the step was rejected, then we need to store the rejected time in the
+            # rejected buffer and decrement the rejected index.
+            i_rjct = jnp.where(keep_step, i_rjct, i_rjct - 1)
+            i_rjct = eqx.error_if(
+                i_rjct,
+                i_rjct < 0,
+                "Maximum number of rejected steps reached. "
+                "Consider increasing JumpStepWrapper.rejected_step_buffer_len.",
+            )
+            rjct_buff = jnp.where(keep_step, rjct_buff, rjct_buff.at[i_rjct].set(t1))
+
+        # Now move on to the NEXT STEP
+        dt_proposal = next_t1 - next_t0
+        # The following line is so that in case prev_dt was intended to be large,
+        # but then clipped to very small (because of step_ts or jump_ts), we don't
+        # want it to stick to very small steps (e.g. the PID controller can only
+        # increase steps by a factor of 10 at a time).
+        dt_proposal = jnp.where(
+            keep_step, jnp.maximum(dt_proposal, prev_dt), dt_proposal
+        )
+        new_prev_dt = dt_proposal
+        next_t1 = next_t0 + dt_proposal
+
+        # If t1 hit a jump point, and the step was kept then we need to set
+        # `next_t0 = nextafter(nextafter(t1))` to ensure that we really skip
+        # over the jump and don't evaluate the vector field at the discontinuity.
+        if jnp.issubdtype(jnp.result_type(next_t0), jnp.inexact):
+            # Two nextafters. If made_jump then t1 = prevbefore(jump location)
+            # so now _t1 = nextafter(jump location)
+            # This is important because we don't know whether or not the jump is as a
+            # result of a left- or right-discontinuity, so we have to skip the jump
+            # location altogether.
+            jump_keep = made_jump & keep_step
+            next_t0 = static_select(
+                jump_keep, eqxi.nextafter(eqxi.nextafter(next_t0)), next_t0
+            )
+
+        if TYPE_CHECKING:
+            assert isinstance(
+                next_t0, RealScalarLike
+            ), f"type(next_t0) = {type(next_t0)}"
```

Comment:
I think we should be able to do this check even outside of type checking?

Reply:
I get this:
```
AssertionError: type(next_t0) = <class 'jax._src.interpreters.partial_eval.DynamicJaxprTracer'>
```

Reply:
Oh, this is a silly known Python bug. You need `isinstance(next_to, typing.get_args(RealScalarLike))`. Python's unions have the wrong `isinstance` behaviour.